### PR TITLE
Workflow execution: Allow binding execution steps to all lifecycle states

### LINKED ISSUE #1078: Allow binding execution steps to all lifecycle states
Make workflow instance stream processor more flexible by encoding a mapping `lifecycle phase -> bpmn step` in the transformed model and the dynamically dispatching events to handlers that implement a step. This is a generalization of the current `BpmnAspect` concept.

### PR DESCRIPTION
**Pull request by ThorbenLindhauer**
closes #1078

### CODE REVIEW

File: broker-core/src/main/java/io/zeebe/broker/workflow/processor/WorkflowInstanceStreamProcessor.java [hunk 1]
```diff
@@ -1239,19 +1239,18 @@ protected void raiseIncident(
...
 
-    private BpmnAspectEventProcessor() {
-      aspectHandlers = new EnumMap<>(BpmnAspect.class);
+    private BpmnStepProcessor() {
+      stepHandlers = new EnumMap<>(BpmnStep.class);
 
-      aspectHandlers.put(BpmnAspect.TAKE_SEQUENCE_FLOW, new TakeSequenceFlowAspectHandler());
-      aspectHandlers.put(BpmnAspect.CONSUME_TOKEN, new ConsumeTokenAspectHandler());
-      aspectHandlers.put(BpmnAspect.EXCLUSIVE_SPLIT, new ExclusiveSplitAspectHandler());
+      stepHandlers.put(BpmnStep.TAKE_SEQUENCE_FLOW, new TakeSequenceFlowAspectHandler());
```
Comment (Zelldon) [inline broker-core/src/main/java/io/zeebe/broker/workflow/processor/WorkflowInstanceStreamProcessor.java]: Maybe we should then rename the handlers as well?

File: broker-core/src/main/java/io/zeebe/broker/workflow/model/transformation/handler/ActivityHandler.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+  public void transform(Activity element, TransformContext context) {
+    final ExecutableWorkflow workflow = context.getCurrentWorkflow();
+    final ExecutableFlowNode activity =
+        workflow.getElementById(element.getId(), ExecutableFlowNode.class);
+
+    bindLifecycle(context, activity);
+  }
+
+  private void bindLifecycle(TransformContext context, final ExecutableFlowNode activity) {
+    activity.bindLifecycleState(
```
Comment (Zelldon) [inline broker-core/src/main/java/io/zeebe/broker/workflow/model/transformation/handler/ActivityHandler.java]: Do we really need the extra abstraction level (BpmnStep)? I mean wouldn't it be simpler to directly add here the corresponding handlers? Say we have an Interface BpmnStep and each class, which implements this interface defines an step. This means you can call handle on each BpmnStep subclass, I think it is no need for the extra Enumeration.

File: broker-core/src/main/java/io/zeebe/broker/workflow/processor/BpmnStepProcessor.java [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+    this.workflowCache = workflowCache;
+
+    // activity
+    stepHandlers.put(BpmnStep.CREATE_JOB, new CreateJobHandler());
+    stepHandlers.put(
+        BpmnStep.APPLY_INPUT_MAPPING,
+        new InputMappingHandler(payloadCache, workflowInstanceIndex, activityInstanceMap));
+    stepHandlers.put(
+        BpmnStep.APPLY_OUTPUT_MAPPING,
+        new OutputMappingHandler(payloadCache, workflowInstanceIndex, activityInstanceMap));
```
Comment (Zelldon) [inline broker-core/src/main/java/io/zeebe/broker/workflow/processor/BpmnStepProcessor.java]: If we have the handle method directly in the step subclass we also have not to do this mapping.

Comment (Zelldon) [reply broker-core/src/main/java/io/zeebe/broker/workflow/processor/BpmnStepProcessor.java]: This is makes it also easier in the future to add more stuff

**Review by ChrisKujawa — CHANGES_REQUESTED** [review]

**Comment by Zelldon** [comment]
We decided to leave as it is now, since we are not sure if the handlers need to bind on the stream processing life cycle. For example to be called on open. Drawback would be that we have to change multiple places if we add/remove/update things.

**Review by ChrisKujawa — APPROVED** [review]

**Pull request merged**